chemicaltoolbox/autodock_vina/docking/docking.xml

```diff
@@ -8,49 +8,58 @@
         <exit_code range="1" />
     </stdio>
     <command><![CDATA[
-        #if $config_params.config_params == 'vals':
-            vina
-                --center_x '$config_params.center_x' 
-                --center_y '$config_params.center_y' 
-                --center_z '$config_params.center_z' 
-                --size_x '$config_params.size_x' 
-                --size_y '$config_params.size_y' 
-                --size_z '$config_params.size_z' 
-                --exhaustiveness '$config_params.exh' 
-                --num_modes 9999 
-                --energy_range 9999 
-                --receptor '$receptor' 
-                --ligand '$ligand' 
-                --out './output1.dat' 
-                --log './output2.dat' 
-                --cpu \${GALAXY_SLOTS:-1}
-                #if $config_params.seed.seed == 'true':
-                    --seed '$config_params.seed.seed_value'
-                #end if
-        #end if
-        #if $config_params.config_params == 'file':
-            vina 
-                --config '$config_params.box' 
-                --receptor '$receptor' 
-                --ligand '$ligand' 
-                --out './output1.dat' 
-                --log './output2.dat'
-                --cpu \${GALAXY_SLOTS:-1}
-                #if $config_params.exh != "":
-                    --exhaustiveness $config_params.exh
-                #end if
-        #end if
-        #if $output_format == 'sdf':
-            && python '$__tool_directory__/convert_pdbqt_to_sdf.py' './output1.dat' '$sdf_output' 
+        #if $ligands.is_of_type("sdf")
+            obabel -isdf '$ligands' -O ligand.pdbqt -m -p $ph_value &&
         #else
-            && mv ./output1.dat '$file_output1'
-            && mv ./output2.dat '$file_output2'
+            ln -s '$ligands' ligand1.pdbqt &&
         #end if
+        rm -rf output &&
+        mkdir output &&
+        for file in ligand*.pdbqt;
```

Comment:
we should use gnu-parallel here to run these in parallel and not sequentially.

Reply:
I need to investigate this more.

Reply:
I modified this to use gnu parallel, but there are some bits I'm not sure about.

1. There are two bits that control the parallelism a). the `-j` option to parallel that tells parallel how may processes to run; b), the `--cpu` option to vina that tells vina how many cores to use. I'm not sure how these should be specified. There's also the `GALAXY_SLOTS` environment variable that seem to want to be used, but seems to always have a value of 1 in my hands, so I'm not sure how this should be used. 
2. The quotation of the execution string is getting pretty scary. Parallel seems to require its command to be single quoted, but galaxy also needs some arguments to be single quoted (double quotes doesn't seem to work for either). To simplify this as much as possible I'm:
2.1. only quoting params that can sensibly contain spaces
2.2 using the `$'abc\'xyz'` pattern of escaping
3. The way params are specified using a box file and directly is a bit inconsistent. When being specified some optional params have hard-coded default values which I feel is not right. Even if these happen to be the same as vina would use as defaults if the param was not specified I don't think they should be provided in the command e.g. params that have default values provided by vina should be optional in galaxy.
4. Because of the inconsistency of how params are handles the 3rd test is a bit messy. It should (assuming the params issue did not exist) generate the same output as the second test, but it does not so I'm currently providing a separate output file to compare against.

@bgruening @simonbray Your thoughts on these welcome.

Reply:
> There are two bits that control the parallelism a). the -j option to parallel that tells parallel how may processes to run; b), the --cpu option to vina that tells vina how many cores to use. I'm not sure how these should be specified. There's also the GALAXY_SLOTS environment variable that seem to want to be used, but seems to always have a value of 1 in my hands, so I'm not sure how this should be used.

Galaxy slots is by default 1. An admin can set the amount of CPU depending on given infrastructure, per tool or depending on input size etc ... so this can be configured by an Admin. Otherwise, it is 1.

So I guess --cpu should be 1 and  -j should be GALAXY_SLOTS?
We could also make this dependent on the amount of molecules. As you prefer Tim. The only thing that we need to take care is that GALAXY_SLOTS is the max amount of CPUs that can be used by this job.

> The quotation of the execution string is getting pretty scary. Parallel seems to require its command to be single quoted, but galaxy also needs some arguments to be single quoted (double quotes doesn't seem to work for either).

Can you not put those commands in a txt file as I did in freebayes?

> he way params are specified using a box file and directly is a bit inconsistent. When being specified some optional params have hard-coded default values which I feel is not right. Even if these happen to be the same as vina would use as defaults if the param was not specified I don't think they should be provided in the command e.g. params that have default values provided by vina should be optional in galaxy.

Then let's make all optional and only pass them to the CMD if given by the user.

Reply:
let's use `--cpu 1` and `-j ` for GALAXY_SLOTS ...